pkg/auth/db_service.go

```diff
@@ -837,3 +868,372 @@ func (s *DBAuthService) markTokenSingleUse(ctx context.Context, tokenID string,
 	}
 	return canUseToken, nil
 }
+
+var ErrExportedEntNotFound = errors.New("previously exported entitiy not found")
+
+func Migrate(ctx context.Context, d *pgxpool.Pool, writer io.Writer) error {
+	je := json.NewEncoder(writer)
+	// Create header
+	if err := je.Encode(kv.Header{
+		LakeFSVersion:   version.Version,
+		PackageName:     model.PackageName,
+		DBSchemaVersion: kv.InitialMigrateVersion,
+		CreatedAt:       time.Now().UTC(),
+	}); err != nil {
+		return err
+	}
+
+	userIDToDetails, err := exportUsers(ctx, d, je)
+	if err != nil {
+		return err
+	}
+
+	groupIDToName, err := exportGroups(ctx, d, je)
+	if err != nil {
+		return err
+	}
+
+	policyIDToName, err := exportPolicies(ctx, d, je)
+	if err != nil {
+		return err
+	}
+
+	if err = exportUserGroups(ctx, d, je, userIDToDetails, groupIDToName); err != nil {
+		return err
+	}
+
+	if err = exportUserPolicies(ctx, d, je, userIDToDetails, policyIDToName); err != nil {
+		return err
+	}
+
+	if err = exportGroupPolicies(ctx, d, je, groupIDToName, policyIDToName); err != nil {
+		return err
+	}
+
+	if err = exportCredentials(ctx, d, je, userIDToDetails); err != nil {
+		return err
+	}
+
+	if err = exportExpiredTokens(ctx, d, je); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+type userDetails struct {
```

Comment:
Move type declarations to the top of the module

Reply:
Done